pulsar/internal/client_handlers.go

```diff
@@ -36,6 +36,13 @@ func (h *ClientHandlers) Add(c Closable) {
 	defer h.l.Unlock()
 	h.handlers[c] = true
 }
+
+func (h *ClientHandlers) Del(c Closable) {
+	h.l.Lock()
+	defer h.l.Unlock()
+	delete(h.handlers, c)
+}
+
```

Comment:
The close method in this class will have to change otherwise there is a deadlock.

We should make a copy of the handlers then release the lock and then call close on the list of handlers. Otherwise the one handler (producer or consumer) will try to call Del() and be deadlocked.

Something like this should work

func (h *ClientHandlers) Close() {
        h.l.Lock()
	handlers := make([]Closable, 0, len(h.handlers))
	for handler := range h.handlers {
		handlers = append(handlers, handler)
	}
	h.l.Unlock()

	for _, handler := range handlers {
		handler.Close()
	}
}